Libraries/LibChess/UCIEndpoint.cpp

```diff
@@ -0,0 +1,132 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "UCIEndpoint.h"
+#include <AK/ByteBuffer.h>
+#include <AK/String.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/File.h>
+
+// #define UCI_DEBUG
+
+namespace Chess::UCI {
+
+Endpoint::Endpoint(NonnullRefPtr<Core::IODevice> in, NonnullRefPtr<Core::IODevice> out)
+    : m_in(in)
+    , m_out(out)
+    , m_in_notifier(Core::Notifier::construct(in->fd(), Core::Notifier::Read))
+{
+    set_in_notifier();
+}
+
+void Endpoint::send_command(const Command& command)
+{
+#ifdef UCI_DEBUG
+    dbg() << class_name() << " Sent UCI Command: " << String(command.to_string().characters(), Chomp);
+#endif
+    m_out->write(command.to_string());
+}
+
+void Endpoint::event(Core::Event& event)
+{
+    switch (event.type()) {
+    case Command::Type::UCI:
+        return handle_uci();
+    case Command::Type::Debug:
+        return handle_debug(static_cast<const DebugCommand&>(event));
+    case Command::Type::IsReady:
+        return handle_uci();
+    case Command::Type::SetOption:
+        return handle_setoption(static_cast<const SetOptionCommand&>(event));
+    case Command::Type::Position:
+        return handle_position(static_cast<const PositionCommand&>(event));
+    case Command::Type::Go:
+        return handle_go(static_cast<const GoCommand&>(event));
+    case Command::Type::Stop:
+        return handle_stop();
+    case Command::Type::Id:
+        return handle_id(static_cast<const IdCommand&>(event));
+    case Command::Type::UCIOk:
+        return handle_uciok();
+    case Command::Type::ReadyOk:
+        return handle_readyok();
+    case Command::Type::BestMove:
+        return handle_bestmove(static_cast<const BestMoveCommand&>(event));
+    case Command::Type::Info:
+        return handle_info(static_cast<const InfoCommand&>(event));
+    default:
+        break;
+    }
+}
+
+void Endpoint::set_in_notifier()
+{
+    m_in_notifier = Core::Notifier::construct(m_in->fd(), Core::Notifier::Read);
+    m_in_notifier->on_ready_to_read = [this] {
+        while (m_in->can_read_line())
+            Core::EventLoop::current().post_event(*this, read_command());
+    };
+}
+
+NonnullOwnPtr<Command> Endpoint::read_command()
+{
+    String line(ReadonlyBytes(m_in->read_line(4096).bytes()), Chomp);
+
+#ifdef UCI_DEBUG
+    dbg() << class_name() << " Recieved UCI Command: " << line;
+#endif
+
+    if (line == "uci") {
+        return make<UCICommand>(UCICommand::from_string(line));
```

Comment:
This (and below).

How about making the Command::from_string return a `NonNullOwnPtr` directly instead ? This way, you don't have to create a temporary object before wrapping it in its pointer.

Also, you could return an `OwnPtr` instead of a `NonNullOwnPtr` to handle invalid input cleanly, but maybe that's not needed (right now, all invalid inputs would assert)

Reply:
> How about making the Command::from_string return a `NonNullOwnPtr` directly instead ? This way, you don't have to create a temporary object before wrapping it in its pointer.

I would consider this a premature optimization, given that for each command you'll spend a second waiting for the engine. Also I think there is a chance that `from_string` could be paired with `send_command`.

> Also, you could return an `OwnPtr` instead of a `NonNullOwnPtr` to handle invalid input cleanly, but maybe that's not needed (right now, all invalid inputs would assert)

Since a chess engine is not user facing, any malformatted commands are a bug in something, and just ignoring them would likely lead to hangs, crashes, illegal moves and all other kinds of shenanigans. Also, because Chess is not particularly critical to system function, I think it is best to assert here.

Reply:
Okay ! I think I misunderstood the purpose of UCI, for some reason I thought it could be used as an online protocol as well. It makes sense for it to ASSERT if it's only for local communications (like IPC currently does).